commcare_export/checkpoint.py

```diff
@@ -0,0 +1,85 @@
+import datetime
+import logging
+import uuid
+
+from commcare_export.writers import SqlMixin
+
+logger = logging.getLogger(__name__)
+
+
+class CheckpointManager(SqlMixin):
+    table_name = 'commcare_export_runs'
+
+    def set_checkpoint(self, query, query_md5, checkpoint_time=None, run_complete=False):
+        logger.info('Setting checkpoint')
+        checkpoint_time = checkpoint_time or datetime.datetime.utcnow()
+        self._insert_checkpoint(
+            id=uuid.uuid4().hex,
+            query_file_name=query,
+            query_file_md5=query_md5,
+            time_of_run=checkpoint_time.isoformat(),
+            final=run_complete
+        )
+        if run_complete:
+            self._cleanup(query_md5)
+
+    def create_checkpoint_table(self):
+        columns = self._get_checkpoint_table_columns()
+        self._migrate_checkpoint_table(columns)
+
+    def _insert_checkpoint(self, **row):
+        table = self.table(self.table_name)
+        insert = table.insert().values(**row)
+        self.connection.execute(insert)
+
+    def _cleanup(self, query_md5):
+        sql = """
+           DELETE FROM {}
+           WHERE final = :final
+           AND query_file_md5 = :md5
+       """.format(self.table_name)
+        self.connection.execute(
+            self.sqlalchemy.sql.text(sql),
+            final=False,
+            md5=query_md5,
+        )
+
+    def _migrate_checkpoint_table(self, columns):
```

Comment:
This is much more strict than the previous 'migration' which was a 'best effort' migration.